### PR TITLE
Binarized Neural Net

### PR DESCRIPTION
**Pull request by volf52**
- Weights (and biases) are initialized as {-1, 1}.
- After each backwards pass, the weights and biases are binarized (to ensure they are binary in the next run, and also when we save them).

- Results for [784, 64, 10] with batchsize 100 and learning rate 0.007: 
Epoch 246: 1094 / 10000 (10.94000%)	Test Loss: 12.51 (same as start)

- Results for [784, 500, 1000, 10] with batchsize 100 and learning rate 0.007: 
Epoch 31: 982 / 10000 (9.82000%)	Test Loss: 45.65 (same as start)

### CODE REVIEW

**Comment by serco425** [comment]
is that12% error? or accuracy ?

**Comment by volf52** [comment]
> is that12% error? or accuracy ?

Loss by the cross_entropy function. Remains the same throughout.

**Comment by volf52** [comment]
I tried to implement some (minor) changes based on what I saw here:

1- [Repo](https://github.com/jaygshah/Binary-Neural-Networks)
2- [Paper](https://arxiv.org/pdf/1602.02830.pdf)    [Github repo for the paper](https://github.com/MatthieuCourbariaux/BinaryNet)

Tried changing the hardtan function. Tried using hard_sigmoid in the binarization part. Tried clipping and then binarizing the weights (as shown in the paper).

TODO: 
- Check the loss function show [here](https://github.com/MatthieuCourbariaux/BinaryNet/blob/d7d45aefd2ada203575b29ebb148fb2c4d925e6d/Train-time/mnist.py#L175)
- Try the training loop without binarizing the biases.

**Comment by volf52** [comment]
> I tried to implement some (minor) changes based on what I saw here:
> 
> 1- [Repo](https://github.com/jaygshah/Binary-Neural-Networks)
> 2- [Paper](https://arxiv.org/pdf/1602.02830.pdf) [Github repo for the paper](https://github.com/MatthieuCourbariaux/BinaryNet)
> 
> Tried changing the hardtan function. Tried using hard_sigmoid in the binarization part. Tried clipping and then binarizing the weights (as shown in the paper).
> 
> TODO:
> 
> * Check the loss function show [here](https://github.com/MatthieuCourbariaux/BinaryNet/blob/d7d45aefd2ada203575b29ebb148fb2c4d925e6d/Train-time/mnist.py#L175)
> * Try the training loop without binarizing the biases.

@serco425 
@indexxy

**Comment by volf52** [comment]
Using square hinge loss ([here](https://github.com/MatthieuCourbariaux/BinaryNet/blob/d7d45aefd2ada203575b29ebb148fb2c4d925e6d/Train-time/mnist.py#L175)) didn't help at all. (The derivative taken from [wikipedia](https://en.wikipedia.org/wiki/Hinge_loss) and [staack exchange](https://stats.stackexchange.com/questions/4608/gradient-of-hinge-loss)).

@serco425 
@indexxy

**Comment by volf52** [comment]
Same for training without binarizing the biases.
